File: examples/cpp/fulfillment/src/fulfillment_logic.cc
#include "fulfillment_logic.hpp"
#include "validation_error.hpp"
#include <chrono>

namespace fulfillment {

using namespace angzarr;

FulfillmentState FulfillmentLogic::rebuild_state(const angzarr::EventBook* event_book) {
    FulfillmentState state;
    if (!event_book || event_book->pages().empty()) return state;

    for (const auto& page : event_book->pages()) {
        if (page.has_event()) {
            state = apply_event(std::move(state), page.event());
        }
    }
    return state;
}

examples::ShipmentCreated FulfillmentLogic::handle_create_shipment(
    const FulfillmentState& state, const std::string& order_id,
    const std::vector<FulfillmentItem>& items) {
    if (state.exists()) throw ValidationError::failed_precondition("Shipment already exists");
    if (order_id.empty()) throw ValidationError::invalid_argument("Order ID is required");
    if (items.empty()) throw ValidationError::invalid_argument("Shipment must have items");

    examples::ShipmentCreated event;
    event.set_order_id(order_id);
    for (const auto& item : items) {
        auto* shipment_item = event.add_items();
        shipment_item->set_product_id(item.product_id);
        shipment_item->set_quantity(item.quantity);
    }
    event.mutable_created_at()->set_seconds(
        std::chrono::system_clock::to_time_t(std::chrono::system_clock::now()));
    return event;
}

examples::ItemsPicked FulfillmentLogic::handle_mark_picked(const FulfillmentState& state) {
    if (!state.exists()) throw ValidationError::failed_precondition("Shipment does not exist");
    if (state.status != FulfillmentStatus::Pending) throw ValidationError::failed_precondition("Shipment not in pending state");

    examples::ItemsPicked event;
    event.mutable_picked_at()->set_seconds(
        std::chrono::system_clock::to_time_t(std::chrono::system_clock::now()));
    return event;
}

examples::ItemsPacked FulfillmentLogic::handle_mark_packed(const FulfillmentState& state) {
    if (!state.exists()) throw ValidationError::failed_precondition("Shipment does not exist");
    if (state.status != FulfillmentStatus::Picking) throw ValidationError::failed_precondition("Items must be picked first");

    examples::ItemsPacked event;
    event.mutable_packed_at()->set_seconds(
        std::chrono::system_clock::to_time_t(std::chrono::system_clock::now()));
    return event;
}

examples::Shipped FulfillmentLogic::handle_ship(
    const FulfillmentState& state, const std::string& tracking_number,
    const std::string& carrier) {
    if (!state.exists()) throw ValidationError::failed_precondition("Shipment does not exist");
    if (state.status != FulfillmentStatus::Packing) throw ValidationError::failed_precondition("Items must be packed first");
    if (tracking_number.empty()) throw ValidationError::invalid_argument("Tracking number is required");
    if (carrier.empty()) throw ValidationError::invalid_argument("Carrier is required");

    examples::Shipped event;
    event.set_tracking_number(tracking_number);
    event.set_carrier(carrier);
    event.mutable_shipped_at()->set_seconds(
        std::chrono::system_clock::to_time_t(std::chrono::system_clock::now()));
    return event;
}

examples::Delivered FulfillmentLogic::handle_record_delivery(const FulfillmentState& state) {
    if (!state.exists()) throw ValidationError::failed_precondition("Shipment does not exist");
    if (state.status != FulfillmentStatus::Shipped) throw ValidationError::failed_precondition("Shipment must be shipped first");

    examples::Delivered event;
    event.mutable_delivered_at()->set_seconds(
        std::chrono::system_clock::to_time_t(std::chrono::system_clock::now()));
    return event;
}

FulfillmentState FulfillmentLogic::apply_event(FulfillmentState state, const google::protobuf::Any& event) {
    const auto& type_url = event.type_url();

    if (type_url.find("ShipmentCreated") != std::string::npos) {
        examples::ShipmentCreated e;
        event.UnpackTo(&e);
        state.order_id = e.order_id();
        state.items.clear();
        for (const auto& item : e.items()) {
            state.items.push_back({item.product_id(), item.quantity()});
        }
        state.status = FulfillmentStatus::Pending;
    } else if (type_url.find("ItemsPicked") != std::string::npos) {
        state.status = FulfillmentStatus::Picking;
    } else if (type_url.find("ItemsPacked") != std::string::npos) {
        state.status = FulfillmentStatus::Packing;
    } else if (type_url.find("Shipped") != std::string::npos) {
        examples::Shipped e;
        event.UnpackTo(&e);
        state.tracking_number = e.tracking_number();
        state.carrier = e.carrier();
        state.status = FulfillmentStatus::Shipped;
    } else if (type_url.find("Delivered") != std::string::npos) {
        state.status = FulfillmentStatus::Delivered;
    }

    return state;
}

}  // namespace fulfillment

File: examples/ruby/fulfillment/lib/fulfillment_logic.rb
# frozen_string_literal: true

require 'time'
require_relative 'command_validation_error'

# Pure business logic for fulfillment aggregate.
# No gRPC dependencies - can be tested in isolation.
module FulfillmentLogic
  # State machine: pending -> picking -> packing -> shipped -> delivered
  FulfillmentState = Struct.new(:order_id, :status, :tracking_number, :carrier,
                                 :shipped_at, :delivered_at, keyword_init: true) do
    def initialize(order_id: '', status: :uninitialized, tracking_number: '',
                   carrier: '', shipped_at: nil, delivered_at: nil)
      super
    end

    def exists?
      status != :uninitialized
    end
  end

  class << self
    def rebuild_state(event_book)
      state = FulfillmentState.new

      return state if event_book.nil? || event_book.pages.empty?

      # Apply events
      event_book.pages.each do |page|
        next unless page.event

        state = apply_event(state, page.event)
      end

      state
    end

    def handle_create_shipment(state, order_id, items)
      raise CommandValidationError.failed_precondition('Shipment already exists') if state.exists?
      raise CommandValidationError.invalid_argument('Order ID is required') if order_id.nil? || order_id.empty?

      Examples::ShipmentCreated.new(
        order_id: order_id,
        items: items || [],
        created_at: now_timestamp
      )
    end

    def handle_mark_picked(state)
      raise CommandValidationError.failed_precondition('Shipment does not exist') unless state.exists?
      raise CommandValidationError.failed_precondition("Cannot pick from status #{state.status}") unless state.status == :pending

      Examples::ItemsPicked.new(picked_at: now_timestamp)
    end

    def handle_mark_packed(state)
      raise CommandValidationError.failed_precondition('Shipment does not exist') unless state.exists?
      raise CommandValidationError.failed_precondition("Cannot pack from status #{state.status}") unless state.status == :picking

      Examples::ItemsPacked.new(packed_at: now_timestamp)
    end

    def handle_ship(state, tracking_number, carrier)
      raise CommandValidationError.failed_precondition('Shipment does not exist') unless state.exists?
      raise CommandValidationError.failed_precondition("Cannot ship from status #{state.status}") unless state.status == :packing
      raise CommandValidationError.invalid_argument('Tracking number is required') if tracking_number.nil? || tracking_number.empty?
      raise CommandValidationError.invalid_argument('Carrier is required') if carrier.nil? || carrier.empty?

      Examples::Shipped.new(
        tracking_number: tracking_number,
        carrier: carrier,
        shipped_at: now_timestamp
      )
    end

    def handle_record_delivery(state)
      raise CommandValidationError.failed_precondition('Shipment does not exist') unless state.exists?
      raise CommandValidationError.failed_precondition("Cannot deliver from status #{state.status}") unless state.status == :shipped

      Examples::Delivered.new(delivered_at: now_timestamp)
    end

    private

    def apply_event(state, event_any)
      type_url = event_any.type_url

      if type_url.end_with?('ShipmentCreated')
        event = Examples::ShipmentCreated.decode(event_any.value)
        FulfillmentState.new(
          order_id: event.order_id,
          status: :pending,
          tracking_number: '',
          carrier: '',
          shipped_at: nil,
          delivered_at: nil
        )
      elsif type_url.end_with?('ItemsPicked')
        FulfillmentState.new(
          order_id: state.order_id,
          status: :picking,
          tracking_number: state.tracking_number,
          carrier: state.carrier,
          shipped_at: state.shipped_at,
          delivered_at: state.delivered_at
        )
      elsif type_url.end_with?('ItemsPacked')
        FulfillmentState.new(
          order_id: state.order_id,
          status: :packing,
          tracking_number: state.tracking_number,
          carrier: state.carrier,
          shipped_at: state.shipped_at,
          delivered_at: state.delivered_at
        )
      elsif type_url.end_with?('Shipped')
        event = Examples::Shipped.decode(event_any.value)
        FulfillmentState.new(
          order_id: state.order_id,
          status: :shipped,
          tracking_number: event.tracking_number,
          carrier: event.carrier,
          shipped_at: event.shipped_at,
          delivered_at: state.delivered_at
        )
      elsif type_url.end_with?('Delivered')
        event = Examples::Delivered.decode(event_any.value)
        FulfillmentState.new(
          order_id: state.order_id,
          status: :delivered,
          tracking_number: state.tracking_number,
          carrier: state.carrier,
          shipped_at: state.shipped_at,
          delivered_at: event.delivered_at
        )
      else
        state
      end
    end

    def now_timestamp
      Google::Protobuf::Timestamp.new(seconds: Time.now.to_i)
    end
  end
end
